internal/pkg/users/users.go

```diff
@@ -17,25 +17,65 @@ limitations under the License.
 package users
 
 import (
+	"bytes"
 	"errors"
+	"fmt"
 	"os/exec"
 	"os/user"
 	"strconv"
-	"strings"
 )
 
-// GetUID returns uid of given username and logs a warning if its missing
-func GetUID(name string) (int, error) {
-	entry, err := user.Lookup(name)
-	if err == nil {
-		return strconv.Atoi(entry.Uid)
-	}
-	if errors.Is(err, user.UnknownUserError(name)) {
+const (
+	// An unknown (i.e. invalid) user ID. It's distinct from a UID's zero value,
+	// which happens to be [RootUID]. Assuming root may or may not be a good
+	// default, depending on the use case. Setting file ownership to root is a
+	// restrictive and safe default, running programs as root is not. Therefore,
+	// this is the preferred return value for UIDs in case of error; callers
+	// must then explicitly decide on the fallback instead of accidentally
+	// assuming root.
+	UnknownUID = -1
+
+	RootUID = 0 // User ID of the root user
+)
+
+var ErrNotExist = errors.New("user does not exist")
+
+// Lookup looks up a user's UID by username. If the user cannot be found, the
+// returned error is [ErrNotExist]. If an error is returned, the returned UID
+// will be [UnknownUID].
+func LookupUID(name string) (int, error) {
+	var uid string
+
+	if entry, err := user.Lookup(name); err != nil {
+		if !errors.Is(err, user.UnknownUserError(name)) {
+			return UnknownUID, err
+		}
+
+		err = ErrNotExist
+
 		// fallback to call external `id` in case NSS is used
-		out, err := exec.Command("/usr/bin/id", "-u", name).CombinedOutput()
-		if err == nil {
-			return strconv.Atoi(strings.TrimSuffix(string(out), "\n"))
+		out, idErr := exec.Command("id", "-u", name).Output()
+		if idErr != nil {
+			var exitErr *exec.ExitError
+			if errors.As(idErr, &exitErr) {
+				firstLine, _, _ := bytes.Cut(exitErr.Stderr, []byte{'\n'})
+				return UnknownUID, fmt.Errorf("%w (%w: %s)", err, idErr, string(firstLine))
+			}
+			return UnknownUID, fmt.Errorf("%w (%w)", err, idErr)
 		}
+
+		uid = string(bytes.TrimSuffix(out, []byte{'\n'}))
```

Comment:
To be precise, this removes a trailing newline, and doesn't return only the first line, but all of them, if any. `id` is expected to only write a single line to stdout, though.

The stderr truncation is different. In case of an error, we never know how many lines will be written. Having only the first line in the error message seemed like a good idea to me, to not potentially pollute the error message but still be helpful. The standard error message of `*exec.ExitError` is not overly helpful on its own, it's usually just "exit status 1" and that's it.

Reply:
PTAL